### PR TITLE
fix(geometry): preserve quaternion rotation matrix dtype

### LINKED ISSUE #3954: quaternion_to_rotation_matrix silently upcasts float16/bfloat16 to float32 (hard-coded torch.tensor(1.0))
`quaternion_to_rotation_matrix` silently upcasts `float16` and `bfloat16` inputs to `float32`. It is the only symbol in the rotation-representation group that does not preserve its input dtype — `normalize_quaternion` and `quaternion_to_axis_angle` on the very same input return `float16`/`bfloat16`.

The cause is a single hard-coded float32 literal (`kornia/geometry/conversions.py:699`):

```python
one: torch.Tensor = torch.tensor(1.0)

matrix_flat: torch.Tensor = torch.stack(
    (
        one - (tyy + tzz),
        ...
    ),
    dim=-1,
)
```

`torch.tensor(1.0)` has no `dtype`/`device` argument, so it is `float32` on cpu; type promotion inside `torch.stack` then lifts the whole matrix to `float32`.

## Repro

```python
import torch
from kornia.geometry.conversions import (
    normalize_quaternion as nq,
    quaternion_to_axis_angle as q2aa,
    quaternion_to_rotation_matrix as q2R,
)

for dt in (torch.float16, torch.bfloat16, torch.float32, torch.float64):
    q = torch.tensor([1.0, 0.0, 0.0, 0.0], dtype=dt)
    print(f"{str(dt):<16} q2R -> {str(q2R(q).dtype):<16} nq -> {str(nq(q).dtype):<16} q2aa -> {q2aa(q).dtype}")
```

## Observed

```
torch.float16    q2R -> torch.float32    nq -> torch.float16    q2aa -> torch.float16
torch.bfloat16   q2R -> torch.float32    nq -> torch.bfloat16   q2aa -> torch.bfloat16
torch.float32    q2R -> torch.float32    nq -> torch.float32    q2aa -> torch.float32
torch.float64    q2R -> torch.float64    nq -> torch.float64    q2aa -> torch.float64
```

`float64` survives only because `float64` wins the promotion against the `float32` literal; the two half-precision dtypes lose it.

## Expected

`quaternion_to_rotation_matrix(q).dtype == q.dtype` for every floating dtype, as every other symbol in this group already guarantees. The one-line fix is to build the literal from the input:

```python
one = torch.tensor(1.0, dtype=quaternion.dtype, device=quaternion.device)
```

or to replace it with `torch.ones_like(tyy)`.

Why it matters beyond tidiness: in a mixed-precision or half-precision pipeline this call silently doubles the memory of the rotation matrix and forces every downstream op into `float32`, either breaking `dtype` assertions or, worse, passing them quietly while defeating the precision policy the user configured. It is also the kind of defect that hides from tests parameterized only over `float32`/`float64`.

## Classification

**Broken output** in the narrow sense that the returned dtype contradicts kornia's uniform dtype-preservation contract, but with essentially zero blast radius: the numeric values are unchanged, only the container dtype differs, and the change affects exactly the two half-precision paths. Per the reference-core lifecycle (Part I / Part III §2) this is an internal repair with no observable default/semantic change on `float32`/`float64` — it can land as an ordinary bugfix under #3916-style byte-identity discipline on those paths.

## Environment

kornia 0.9.0rc1 @ `3358ef16`, torch 2.9.1, cpu. Executed from a worktree carrying the batch-3a docs/tests commits on top of `3358ef16`; `quaternion_to_rotation_matrix` is byte-identical to `origin/main` there (no diff hunk lands in it; function body compares AST-identical).

Posted on behalf of @ducha-aiki by Claude (Fable 5).

### PR DESCRIPTION
**Pull request by fallenmi**
## Which lane? *(check one — see [Contributing Guide](https://github.com/kornia/kornia/blob/main/CONTRIBUTING.md#which-lane-is-your-contribution))*
- [x] 🟢 **Green lane** — test-first bug fix / verified docs fix / [`help wanted`](https://github.com/kornia/kornia/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22help%20wanted%22) issue / benchmark results. No prior issue needed.
- [ ] 🟠 **Discuss-first** — feature or behavior change; a maintainer confirmed the scope on the linked issue.

**Fixes/Relates to:** Fixes #3954

## What does this PR do?

For an unbatched `(4,)` quaternion, each component is a 0-dimensional tensor. The existing 0-dimensional CPU `float32` tensor used for the matrix diagonal therefore ties with a `float16` or `bfloat16` component during type promotion and lifts the result to `float32`.

This replaces that tensor with a Python scalar. PyTorch's wrapped-scalar promotion preserves the input tensor's dtype without allocating a separate device tensor. A focused regression covers both affected low-precision dtypes and uses the identity rotation matrix as its value oracle.

This branch is rebased onto current `main` (`d73bf9c0`). Merged PR #3967 added temporary #3954 documentation, a strict xfail/wart pair, and `_undo_3954_upcast` casts. The fix makes the xfail XPASS and the wart fail for both dtypes, so this PR also retires only that now-obsolete #3954 scaffolding. All unrelated convention and wart coverage remains and now checks the function output directly.

## Test log

Environment: macOS arm64, Python 3.13.5, PyTorch 2.13.0, CPU.

Current `main` baseline:

```text
$ pytest -q \
    ...::test_convention_output_dtype_equals_input_dtype_3954 \
    ...::test_wart_half_precision_input_is_upcast_to_float32_3954 \
    --device=cpu
2 passed, 2 xfailed in 0.43s
```

With the functional fix applied before cleanup, the two strict-xfail cells become `XPASS(strict)` failures and both wart cells fail because the output is now `float16`/`bfloat16`. This execution proves those two temporary pins are obsolete rather than merely stale-looking.

Current rebased head:

```text
$ pytest -q \
    tests/geometry/test_conversions.py::TestQuaternionToRotationMatrix::test_unbatched_preserves_low_precision_dtype \
    --device=cpu
2 passed in 0.33s

$ KORNIA_TEST_DTYPE=float16,bfloat16 pytest -q \
    tests/geometry/test_conversions.py::TestQuaternionToRotationMatrix \
    --device=cpu
34 passed, 4 skipped, 1 xfailed in 0.35s

$ KORNIA_TEST_DTYPE=float16,bfloat16 pytest -q \
    tests/geometry/test_conversions.py::TestAngleAxisToRotationMatrix::test_convention_positive_angle_about_z_maps_x_to_y \
    tests/geometry/test_conversions.py::TestQuaternionFromEuler::test_convention_composition_is_rz_ry_rx \
    --device=cpu
4 passed in 0.27s

$ pytest -q tests/geometry/test_conversions.py tests/integration/test_conversions.py --device=cpu
487 passed, 1 skipped, 24 xfailed in 1.51s

$ ruff check kornia/geometry/conversions.py tests/geometry/test_conversions.py
All checks passed!

$ ruff format --check kornia/geometry/conversions.py tests/geometry/test_conversions.py
2 files already formatted

$ uvx --from 'ty==0.0.9' ty check kornia
All checks passed!

$ git diff --check origin/main...HEAD
# no output
```

The listed xfails are unrelated current-main convention pins; the #3954 strict xfail and wart were removed only after the execution proof above.

## AI Usage Disclosure *(pick the closest — the 🟡/🔴 line is fuzzy and only deception is sanctioned; none of these is a bad answer, see [AI_POLICY.md](https://github.com/kornia/kornia/blob/main/AI_POLICY.md))*
- [ ] 🟢 **Human-written**
- [ ] 🟡 **AI-assisted** — AI helped; I reviewed the resulting change and ran the relevant tests
- [x] 🔴 **AI-generated** — an agent wrote most of the change; automated verification is documented above, and personal human review is still pending

Generated-By: OpenAI Codex (GPT-5). This remains a Draft pending personal human line-by-line review before readiness.

### CODE REVIEW

**Review by ducha-aiki — DISMISSED** [review]
Correct fix, and the diff is smaller than the defect it closes. Verified on current `main` (`35555b55`), not on your branch — see the rebase note.

## Verified

```text
                    main                 with this PR
float16   (4,)   -> torch.float32     -> torch.float16
bfloat16  (4,)   -> torch.float32     -> torch.bfloat16
float32   (4,)   -> torch.float32     -> torch.float32
float64   (4,)   -> torch.float64     -> torch.float64
float16 (1, 4)   -> torch.float16     -> torch.float16   (already correct, unchanged)
```

`float32` and `float64` outputs are **bitwise identical** before and after — I compared the raw words via `.view(torch.int32)` / `.view(torch.int64)` on a batch of five random unit quaternions and got the same integers on both trees. Replacing a 0-dim `torch.tensor(1.0)` with a Python `1.0` turns a participant in type promotion into a wrapped scalar, which is exactly the right lever: it cannot move a result, only a dtype.

`grep -rnE "#3954|issues/3954" kornia/` after your change: no hits.

## Your change fixes 16 half-precision test failures you did not claim

`tests/geometry --dtype=float16,bfloat16` on current `main`: **696 failures**. With your diff applied: **680**, and the diff of the two ID sets is 16 removals and **zero additions**:

```text
TestQuaternionToRotationMatrix::test_unit_quaternion   [float16, bfloat16]
TestQuaternionToRotationMatrix::test_x_rotation        [float16, bfloat16]
TestQuaternionToRotationMatrix::test_y_rotation        [float16, bfloat16]
TestQuaternionToRotationMatrix::test_z_rotation        [float16, bfloat16]
TestRotationMatrixToQuaternion::test_back_and_forth    [float16, bfloat16]
TestRotationMatrixToQuaternion::test_cond1_180_rot_x   [float16, bfloat16]
TestRotationMatrixToQuaternion::test_cond2_180_rot_y   [float16, bfloat16]
tests/geometry/test_quaternion.py::TestQuaternion::test_matrix [float16, bfloat16]
```

Every one of those is an ordinary test comparing against a half-precision expected tensor, failing purely because the actual came back `float32`. This is worth putting in the PR body — "closes `#3954` and repairs 16 pre-existing half-precision failures, with no new ones" is a much stronger statement than the one you made, and it is the kind of number `#4033` (the half-precision CI job) will want on record.

At float32 the suite is unchanged: `tests/geometry` gives 0 failures with your diff, same as `main`.

## One request: keep the issue number on the guard

You replaced `test_convention_output_dtype_equals_input_dtype_3954` with `test_unbatched_preserves_low_precision_dtype` and it is a genuine pin — it fails on unpatched `main` for both half dtypes, and adding the value assertion against `eye(3)` is an improvement over the dtype-only original. Deleting `_undo_3954_upcast` and its four call sites was also exactly right; the helper's own comment asked for it.

But after your change `grep -rnE "#3954" tests/` returns **nothing**. Per `AGENTS.md` a pin goes on naming its issue after the fix lands, and the convention-block program greps by both the `#NNNN` anchor and the `test_convention_*` prefix. As it stands, nothing connects the surviving guard to the closed issue, and the mechanism — 0-dim tensors losing to dimensioned ones in type promotion, so only the unbatched shape was affected — is now recorded nowhere. That mechanism is the non-obvious part; it is why `(4,)` broke and `(1, 4)` did not.

The minimal change I would like: keep the name `test_convention_output_dtype_equals_input_dtype_3954` (just drop the `xfail` marker rather than deleting the test), fold your `assert_close` against `eye(3)` into it, and keep a two-line version of the promotion explanation. That is fewer edits than what you have and it preserves both greps.

## Rebase

Base is `d73bf9c0`, so this branch has not run the pins from `#4006` or `#4028`; CI dates to 2026-08-25. Everything above was measured with your diff applied to current `main`, so I know it is clean — please rebase so your own CI shows it.

Approving on the verification. Restore the `_3954` name on the guard and this is ready.

---
*Posted on behalf of @ducha-aiki by Claude (Opus 5).*

**Review by ducha-aiki — APPROVED** [review]
Pushed the naming change I asked for rather than leaving you a round-trip — head is `25eee7d0`, one commit on top of yours, no rebase needed.

Your replacement guard was correct; it was only unnamed. After the rename `grep -rnE "#3954" tests/` returned nothing, so nothing linked the surviving pin to the closed issue and the mechanism was recorded nowhere. The commit restores the `test_convention_output_dtype_equals_input_dtype_3954` name the convention program greps for and writes the mechanism down: 0-dim tensors lose to dimensioned ones in type promotion, which is exactly why `(4,)` broke and `(1, 4)` did not.

Three small additions while I was there:

- Your `assert_close` against `eye(3)` is kept — it is an improvement on the dtype-only original.
- `_skip_if_dtype_unavailable(device, half_dtype)` is back, so a backend without float16/bfloat16 reports as **skipped** instead of failing on a `RuntimeError`.
- The batched shape is carried as a third assertion, so a fix that special-cased one shape would not satisfy the pin.

Still **2 failed** on the pre-fix library and **2 passed** with your change. `tests/geometry` 2263 passed / 0 failed on the pushed head; `ruff` clean.

Re-approving; the earlier approval was dismissed by my push. Worth restating what this PR actually buys, because your description undersells it: `torch.tensor(1.0)` -> `1.0` leaves float32/float64 **bitwise** identical and repairs **16 pre-existing half-precision failures with zero new ones**.

---
*Posted on behalf of @ducha-aiki by Claude (Opus 5).*

**Review by ducha-aiki — APPROVED** [review]
Final verification round, `25eee7d0` merged onto current `main` (the PR base was 9 commits behind).

- `grep -rnE "#3954|issues/3954" kornia/ docs/` is empty; the two remaining hits are inside the restored pin itself. `_undo_3954_upcast` has no callers left anywhere in the tree.
- `tests/geometry/test_conversions.py --dtype=float32,float64`: 602 passed, 20 xfailed. Doctests pass; ruff clean.
- The restored `test_convention_output_dtype_equals_input_dtype_3954` keeps the name, so the `_3954` anchor the AGENTS.md grep protocol relies on survives.

Worth recording that this does more than close #3954 — it clears **seven** pre-existing bfloat16 failures and introduces none. Diffing the failure sets of `tests/geometry/test_conversions.py --dtype=bfloat16 -k "RotationMatrix or Euler"`:

```
fixed by this PR (fail on main, pass here):
  TestQuaternionToRotationMatrix::test_unit_quaternion
  TestQuaternionToRotationMatrix::test_x_rotation
  TestQuaternionToRotationMatrix::test_y_rotation
  TestQuaternionToRotationMatrix::test_z_rotation
  TestRotationMatrixToQuaternion::test_back_and_forth
  TestRotationMatrixToQuaternion::test_cond1_180_rot_x
  TestRotationMatrixToQuaternion::test_cond2_180_rot_y
new failures introduced: none
```

The seven that still fail at bfloat16 fail identically on `main` and are a separate tolerance question.

Approving. This is still marked **draft** — it needs `gh pr ready 3982` before it can be merged.

*Posted on behalf of @ducha-aiki by Claude (Opus 5).*
